### ISSUE TITLE
CPU Usage seems wrong

### ISSUE THREAD

**Issue by damm** [influxdata/telegraf#1032]
Trying 0.12 of Telegraf and Nightly to ship CPU Usage Metrics I seem to be getting the wrong data.
---
#### telegraf.conf

```
[agent]
flush_interval = "10s"
flush_jitter = "5s"
hostname = "dc-origin02"
interval = "10s"
round_interval = true
[tags]
```
---
#### telegraf.d/default_inputs.conf

```
[inputs.cpu]
drop = ["time_*"]
percpu = false
totalcpu = true
[inputs.disk]
ignore_fs = ["tmpfs", "devtmpfs"]
[inputs.io]
[inputs.mem]
[inputs.net]
[inputs.netstat]
[inputs.swap]
[inputs.system]
```

I am looking at cpu_total here; and not just per cpu.  

```
SELECT max("usage_system") AS "system", max("usage_user") AS "user", max("usage_iowait") AS "iowait", max("usage_irq") AS "irq", max("usage_nice") AS "nice", max("usage_softirq") AS "softirq", max("usage_steal") AS "steal" FROM "cpu" WHERE "host" =~ /^$servers$/ AND "cpu" = 'cpu-total' AND $timeFilter GROUP BY time($interval), "host" fill(none)
```

I'll try and post more information as I can come up with it @sparcc

**Comment by damm** [comment]
Unfortunately I don't see anything obvious I tried using the nightly version and it's the same; it seems to be writing with the precision of "s" which is what I expect.

Logs are quiet which is what I expect too

```
2016/04/15 00:05:20 Gathered metrics, (10s interval), from 11 inputs in 28.783788ms
2016/04/15 00:05:26 Wrote 30 metrics to output influxdb in 20.605512ms
2016/04/15 00:05:30 Gathered metrics, (10s interval), from 11 inputs in 24.368319ms
2016/04/15 00:05:39 Wrote 30 metrics to output influxdb in 358.602492ms
2016/04/15 00:05:40 Gathered metrics, (10s interval), from 11 inputs in 25.579336ms
2016/04/15 00:05:50 Gathered metrics, (10s interval), from 11 inputs in 26.511136ms
2016/04/15 00:05:52 Wrote 60 metrics to output influxdb in 20.448093ms
2016/04/15 00:06:00 Gathered metrics, (10s interval), from 11 inputs in 29.622276ms
2016/04/15 00:06:05 Wrote 30 metrics to output influxdb in 16.43309ms
2016/04/15 00:06:10 Gathered metrics, (10s interval), from 11 inputs in 26.611493ms
2016/04/15 00:06:17 Wrote 30 metrics to output influxdb in 16.974835ms
```

**Comment by damm** [comment]
```
telegraf -config ./telegraf.conf -input-filter cpu:system -test -config-directory /etc/telegraf/telegraf.d/ && w
* Plugin: system, Collection 1
> system load1=0.7,load15=1.12,load5=1.19,n_users=2i,uptime=16679710i,uptime_format="193 days,  1:15" 1460679175965865501
* Plugin: cpu, Collection 1
* Plugin: cpu, Collection 2
> cpu,cpu=cpu-total usage_guest=0,usage_guest_nice=0,usage_idle=95.79124578955648,usage_iowait=0.33670033714905745,usage_irq=0,usage_nice=0,usage_softirq=0,usage_steal=0,usage_system=0,usage_user=3.8720538811338687 1460679176466495742
 00:12:56 up 193 days,  1:15,  2 users,  load average: 0.70, 1.19, 1.12
```

**Comment by damm** [comment]
```
telegraf -config ./telegraf.conf -input-filter cpu:system -test -config-directory /etc/telegraf/telegraf.d/ && w && sar -u 2 5
* Plugin: cpu, Collection 1
* Plugin: cpu, Collection 2
> cpu,cpu=cpu-total usage_guest=0,usage_guest_nice=0,usage_idle=85.71428565998119,usage_iowait=0,usage_irq=0,usage_nice=0,usage_softirq=0,usage_steal=0,usage_system=2.0408163275003357,usage_user=12.244897965002016 1460679339429582821
* Plugin: system, Collection 1
> system load1=0.52,load15=0.37,load5=0.41,n_users=1i,uptime=14082463i,uptime_format="162 days, 23:47" 1460679339430481969
 00:15:39 up 162 days, 23:47,  1 user,  load average: 0.52, 0.41, 0.37
USER     TTY      FROM             LOGIN@   IDLE   JCPU   PCPU WHAT
ubuntu   pts/1    74-94-68-194-bel 00:14    1.00s  0.07s  0.01s sshd: ubuntu [priv]
Linux 3.13.0-66-generic (ip-172-31-49-35) 	04/15/2016 	_x86_64_	(2 CPU)

12:15:39 AM     CPU     %user     %nice   %system   %iowait    %steal     %idle
12:15:41 AM     all     19.02      0.00      2.31      0.00      0.00     78.66
12:15:43 AM     all      6.12      0.00      2.30      0.00      0.00     91.58
12:15:45 AM     all      3.83      0.00      2.55      0.00      0.26     93.37
12:15:47 AM     all      3.08      0.00      2.05      0.00      0.00     94.87
12:15:49 AM     all      2.84      0.00      1.29      0.00      0.00     95.88
Average:        all      6.97      0.00      2.10      0.00      0.05     90.88
```

**Comment by damm** [comment]
@sparrc they seem close actually I do use jitter in my output so it randomizes when it pushes out to influxdb; I dont' suspect this is related?

**Comment by damm** [comment]
![screen shot 2016-04-14 at 5 22 52 pm](https://cloud.githubusercontent.com/assets/1010/14547725/944cd6e4-0265-11e6-9d6b-b5cae4b62b07.png)
![screen shot 2016-04-14 at 5 23 00 pm](https://cloud.githubusercontent.com/assets/1010/14547728/9660d5d4-0265-11e6-934b-8f0d5dc00b00.png)

**Comment by damm** [comment]
Some pictures of it rendered @sparrc

**Comment by sparrc** [comment]
what am I looking at in this graph? is there some relevant information that shows the cpu usage off?

**Comment by damm** [comment]
@sparrc the first graph is taken with Telegraf; the 2nd is using a plugin in Sensu that collects cpu usage.

The difference in CPU Usage between A and B is very large; I'm trying to figure out why it works fine in some installations and here the CPU usage is way off from reality.

It would report the CPU User usage at 11% when infact it was closer to 50% 

If there's a better way to provide test data to prove the problem here; please enlighten me.

P.S. some of those comments are using telegraf -test reporting the results and using afterwards to return some results.  Unfortunately i'm not running them at the same time as it makes it harder to read.  Couldn't think of a better tool other than mpstat which does average the results returned.

**Comment by damm** [comment]
I was reviewing gopsutil and noticed it seems to have cases for 2.6.34+ and 3.2+

This is Ubuntu LTS 14.04 running 3.13.0-65 (a little dated) could that be related?

https://github.com/shirou/gopsutil/blob/5daf090513cd856198b04a3f65dd7539c7602112/cpu/cpu_linux.go#L221

Hosts where it works right are 4.5 (or 4.0+)

**Comment by sparrc** [comment]
Those cases aren't related to this. Some kernels just don't have the fields in /proc/stat to get that information.

**Comment by damm** [comment]
@sparrc so what can I do to help identify the issue here? I know Telegraf works so I would like to get this resolved.

**Comment by sparrc** [comment]
I'm not really sure, if you had some more definitive and specific proof that would help. At the moment it's a bit hard to believe that none of our thousands of users have reported this if it is in fact wrong. The implementation of the CPU plugin hasn't changed since launch.

**Comment by damm** [comment]
Pretty much all I can give you is select statements and point you to the Sensu extension that I collect with.  It's not that much

I attached a few snapshots; the first snapshot is using Telegraf and the second is using the extension (that is partially pasted below).

P.S. They are from the same time frame.

Thanks.

```ruby
      def parse_proc_stat
        sample = {}
        cpu_metrics = %w(user nice system idle iowait irq softirq steal guest)
        misc_metrics = %w(ctxt processes procs_running procs_blocked btime intr)
        read_file('/proc/stat') do |proc_stat|
          proc_stat.each_line do |line|
            next if line.empty?
            data = line.split(/\s+/)
            key = data.shift
            values = data.map(&:to_i)
            if key =~ /cpu([0-9]+|)/
              sample[key] = Hash[cpu_metrics.zip(values)]
              sample[key]['total'] = values.inject(:+)
            elsif misc_metrics.include?(key)
              sample[key] = values.last
            end
          end
          yield sample
        end
      end
```

**Comment by sparrc** [comment]
from your `sar` output above, it looks like telegraf is closer to the real usage percent. I can't really comment on that ruby code it looks like it's just reading the raw time values from /proc/stat

**Comment by damm** [comment]
@sparrc indeed it is reporting the raw values.  It's old but it works; I know mpstat does average the cpu usage some (which would be more acceptable)

I could agree that this is acceptable however like earlier (I wasn't logged into the machine so I could only select the data) however like in this graph; it shows telegraf reporting a iowait of 11.
![screen shot 2016-04-19 at 11 12 50 am](https://cloud.githubusercontent.com/assets/1010/14650247/c7fa637c-061f-11e6-9e83-c11e386f9212.png)

![screen shot 2016-04-19 at 11 12 50 am](https://cloud.githubusercontent.com/assets/1010/14650260/dea8ad36-061f-11e6-8eee-d9799739b0dc.png)

It actually almost hit 100% iowait 

Here's a full copy of the script here https://github.com/sensu/sensu-community-plugins/blob/d7f987a23699bf18cb923ed68a4a434f035751d2/extensions/checks/system_profile.rb

I was hoping to get rid of it as Sensu no longer ships it I believe; as well as moving to Telegraf because the metrics are written correctly (and I don't have to fix plugins).

**Comment by damm** [comment]
I should mention that these servers have 4+ Cores; sometimes up to 24.  We use the hpet clocksource and hpet is enabled on the BIOS. (if it's metal)

Deployed using https://github.com/NorthPage/telegraf-cookbook on Ubuntu 14.04

**Comment by damm** [comment]
3.14 but sometimes it's 2.95x

**Comment by sparrc** [comment]
@damm, my point is this, from your own output above:

telegraf reports 85% idle and 2% system on the test run:

```
> cpu,cpu=cpu-total usage_guest=0,usage_guest_nice=0,usage_idle=85.71428565998119,usage_iowait=0,usage_irq=0,usage_nice=0,usage_softirq=0,usage_steal=0,usage_system=2.0408163275003357,usage_user=12.244897965002016 1460679339429582821
```

sar reports 90% idle and 2% system around the same time:

```
12:15:39 AM     CPU     %user     %nice   %system   %iowait    %steal     %idle
Average:        all      6.97      0.00      2.10      0.00      0.05     90.88
```

so to me this evidence is more convincing that your sensu plugin is wrong. I wouldn't expect these to be exactly the same but I would expect it to be close.

**Comment by damm** [comment]
@sparrc sar averages the results it returns so it's .. additionally I ran those in a sequence.  So they are not returning even the same time span.

Not very conclusive.  I've tried a few other methods using exec plugins to grab the cpu usage but I haven't found a method that works right.

Every other plugin works right; even loadavg and memory is .. Would love to know why the old legacy method works ..

I'll spend a bit more time extracting that method into a exec plugin and give it a whirl; it's strange because it's roughly the same code that is referenced by Telegraf.

You'd think they'd be the same results...

**Comment by sparrc** [comment]
any updates @damm?

**Comment by damm** [comment]
No not really; I suspect it's related to the HZ of the servers in question.

CONFIG_HZ_250=y is standard for Ubuntu; other people set it differently.  That doesn't explain the difference between AWS instances difference vs servers that are not virtualized.

Unless the AWS instance is loosing jiffies.

I should retry after https://github.com/influxdata/telegraf/pull/1234 was merged.  I'll get back to you.

**Comment by sparrc** [comment]
closing this out for lack of updates & corroboration, @damm please feel free to reopen if you're able to reproduce with a more recent version of telegraf

**Comment by damm** [comment]
@sparrc oh I can reproduce it but after a point with 40 core's Telegraf is close enough.

I think the blame is partially due to not all kernels being in the same HZ and I don't think that should complicate calculating cpu usage but it seems to.

Pretty much my tools tell me one thing but i'm not sure I should trust them; i'm dealing with stock ubuntu kernels with hpet enabled in the bios and in the kernel using hpet timing.  I don't think that should matter but my old box i built by hand in 2006 uses telegraf and it works perfect.

no hpet; HZ=1000 

Ubuntu seems to use HZ=250

**Comment by sparrc** [comment]
@damm https://github.com/shirou/gopsutil is the library we use to get CPU info, and it does make corrections for the different hz of various kernels. I would be very surprised if it had a kernel that ubuntu is using wrong, but if you could trace it down I would recommend opening an issue there.

**Comment by damm** [comment]
He also tries to take into cpu frequency scaling (power management) which we disable... well we can't fully disable it on the cpu itself 

Nothing is perfect and I suspect the tooling (top, ps, vmstat) can also be part of the problem as I tried various ways to generate the average (ended up settling with sar) which even gave worse results.

So like I said in the previous post without running SystemTap to see perfectly it's going to be difficult to prove it wrong (or consequentially right).  Some of the  servers I have are 18 core some are 40 core and wee you see 87% cpu usage on a single core on 40-core system and you think it's a lot and whoops 

So could be partially end user too

**Comment by pumantis** [comment]
We are also seeing similar issues.
We're having problems on KVM hypervisors. The values reported by telegraf don't match the ones reported by sar. I attached the output from both on the same time period. In the middle of the period I generated some load on the guests of the hypervisor, that's when the values differ. Sar doesn't report the guest usage values, but the idle and user percentage still should be the same.
[sar.txt](https://github.com/influxdata/telegraf/files/599431/sar.txt)
[telegraf.txt](https://github.com/influxdata/telegraf/files/599432/telegraf.txt)

We have no issues on the guests or physical hosts, only the hypervisors.
